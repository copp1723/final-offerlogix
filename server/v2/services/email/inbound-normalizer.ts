/**
 * Inbound Email Normalizer
 * 
 * Normalizes Mailgun webhook payloads into consistent InboundEmail format.
 * Handles both JSON fields and MIME header parsing.
 */

import type { InboundEmail } from './types';

// ============================================================================
// MAILGUN PAYLOAD INTERFACES
// ============================================================================

interface MailgunWebhookBody {
  // Core fields (always present)
  recipient: string;              // "riley@kunesmacomb.kunesauto.vip"
  sender: string;                 // "lead@example.com"
  subject: string;                // Email subject
  
  // Content fields
  'body-plain'?: string;          // Plain text body
  'body-html'?: string;           // HTML body
  
  // Headers (two possible formats)
  'message-headers'?: Array<[string, string]>; // [["Message-Id", "value"], ...]
  'Message-Id'?: string;          // Direct field access
  'In-Reply-To'?: string;         // Direct field access
  'References'?: string;          // Direct field access
  
  // Additional fields from Mailgun
  'stripped-text'?: string;       // Text with signatures removed
  'stripped-html'?: string;       // HTML with signatures removed
}

// ============================================================================
// EMAIL ADDRESS PARSING
// ============================================================================

/**
 * Parse email address from various formats
 * Handles: "user@domain.com", "Name <user@domain.com>", etc.
 */
function parseEmailAddress(input: string): { email: string; name?: string } {
  if (!input || typeof input !== 'string') {
    throw new Error('Invalid email input');
  }

  const trimmed = input.trim();
  
  // Format: "Name <user@domain.com>"
  const nameEmailMatch = trimmed.match(/^(.+?)\s*<([^>]+)>$/);
  if (nameEmailMatch) {
    const [, name, email] = nameEmailMatch;
    return {
      email: email.toLowerCase().trim(),
      name: name.trim().replace(/^["']|["']$/g, '') // Remove quotes
    };
  }
  
  // Format: "user@domain.com"
  const emailMatch = trimmed.match(/^([^@\s]+@[^@\s]+\.[^@\s]+)$/);
  if (emailMatch) {
    return {
      email: emailMatch[1].toLowerCase().trim()
    };
  }
  
  throw new Error(`Invalid email format: ${input}`);
}

/**
 * Parse domain and local part from email address
 */
function parseEmailParts(email: string): { localPart: string; domain: string } {
  const normalized = email.toLowerCase().trim();
  const atIndex = normalized.lastIndexOf('@');
  
  if (atIndex === -1) {
    throw new Error(`Invalid email format: ${email}`);
  }
  
  return {
    localPart: normalized.substring(0, atIndex),
    domain: normalized.substring(atIndex + 1)
  };
}

// ============================================================================
// HEADER PARSING
// ============================================================================

/**
 * Extract headers from Mailgun payload (handles both formats)
 */
function extractHeaders(body: MailgunWebhookBody): Record<string, string> {
  const headers: Record<string, string> = {};
  
  // Format 1: message-headers array
  if (body['message-headers']) {
    for (const [key, value] of body['message-headers']) {
      if (typeof key === 'string' && typeof value === 'string') {
        headers[key.toLowerCase()] = value;
      }
    }
  }
  
  // Format 2: Direct field access (override array values)
  const directHeaders = [
    'Message-Id',
    'In-Reply-To', 
    'References',
    'Date',
    'Content-Type'
  ];
  
  for (const headerName of directHeaders) {
    const value = body[headerName as keyof MailgunWebhookBody];
    if (typeof value === 'string') {
      headers[headerName.toLowerCase()] = value;
    }
  }
  
  return headers;
}

/**
 * Parse References header into array
 */
function parseReferences(referencesHeader?: string): string[] {
  if (!referencesHeader || typeof referencesHeader !== 'string') {
    return [];
  }
  
  // References format: "<msg1@domain> <msg2@domain> <msg3@domain>"
  const references = referencesHeader
    .trim()
    .split(/\s+/)
    .map(ref => ref.trim())
    .filter(ref => ref.length > 0 && ref.includes('@'))
    .map(ref => {
      // Ensure proper <> wrapping
      if (!ref.startsWith('<') || !ref.endsWith('>')) {
        return ref.includes('<') && ref.includes('>') ? ref : `<${ref}>`;
      }
      return ref;
    });
  
  return references;
}

/**
 * Clean and validate Message-ID format
 */
function normalizeMessageId(messageId?: string): string {
  if (!messageId || typeof messageId !== 'string') {
    throw new Error('Missing Message-ID header');
  }
  
  let cleaned = messageId.trim();
  
  // Ensure proper <> wrapping
  if (!cleaned.startsWith('<') || !cleaned.endsWith('>')) {
    cleaned = `<${cleaned}>`;
  }
  
  // Basic validation
  if (!cleaned.includes('@')) {
    throw new Error(`Invalid Message-ID format: ${cleaned}`);
  }
  
  return cleaned;
}

// ============================================================================
// MAIN NORMALIZER
// ============================================================================

/**
 * Normalize Mailgun webhook payload into InboundEmail
 */
export function normalizeMailgun(body: any): InboundEmail {
  if (!body || typeof body !== 'object') {
    throw new Error('Invalid webhook body');
  }
  
  const mailgunBody = body as MailgunWebhookBody;
  
  // Extract required fields - handle both formats (recipient/sender OR to/from)
  const recipient = mailgunBody.recipient || (mailgunBody as any).to;
  const sender = mailgunBody.sender || (mailgunBody as any).from;
  const subject = mailgunBody.subject;
  
  if (!recipient || !sender || !subject) {
    throw new Error('Missing required fields: recipient/to, sender/from, or subject');
  }
  
  try {
    // Parse recipient (agent) email
    const recipientInfo = parseEmailAddress(mailgunBody.recipient);
    const agentParts = parseEmailParts(recipientInfo.email);
    
    // Parse sender (lead) email  
    const senderInfo = parseEmailAddress(mailgunBody.sender);
    
    // Extract headers
    const headers = extractHeaders(mailgunBody);
    
    // Get Message-ID (required)
    const messageId = normalizeMessageId(headers['message-id'] || mailgunBody['Message-Id']);
    
    // Parse In-Reply-To
    const inReplyToRaw = headers['in-reply-to'] || mailgunBody['In-Reply-To'];
    const inReplyTo = inReplyToRaw ? normalizeMessageId(inReplyToRaw) : null;
    
    // Parse References chain
    const referencesRaw = headers['references'] || mailgunBody['References'];
    const references = parseReferences(referencesRaw);
    
    // If in-reply-to missing but References present, set inReplyTo = last(refs) for resilience
    let finalInReplyTo = inReplyTo;
    if (!finalInReplyTo && references.length > 0) {
      finalInReplyTo = references[references.length - 1];
    }
    
    // Choose best content (prefer stripped versions, fall back to full)
    const text = mailgunBody['stripped-text'] || mailgunBody['body-plain'];
    const html = mailgunBody['stripped-html'] || mailgunBody['body-html'];
    
    if (!text && !html) {
      throw new Error('No email content found (text or html)');
    }
    
    // Build normalized result
    const normalized: InboundEmail = {
      agentLocalPart: agentParts.localPart,
      agentDomain: agentParts.domain,
      fromEmail: senderInfo.email, // Already lowercased in parseEmailAddress
      subject: mailgunBody.subject.trim(),
      text: text?.trim(),
      html: html?.trim(), 
      messageId,
      inReplyTo: finalInReplyTo,
      references,
      rawHeaders: headers
    };
    
    return normalized;
    
  } catch (error) {
    throw new Error(
      `Failed to normalize Mailgun payload: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

// ============================================================================
// VALIDATION HELPERS
// ============================================================================

/**
 * Validate normalized email for completeness
 */
export function validateInboundEmail(email: InboundEmail): void {
  const required = ['agentLocalPart', 'agentDomain', 'fromEmail', 'subject', 'messageId'];
  
  for (const field of required) {
    const value = email[field as keyof InboundEmail];
    if (!value || typeof value !== 'string' || !value.trim()) {
      throw new Error(`Invalid or missing field: ${field}`);
    }
  }
  
  // Must have either text or html content
  if (!email.text && !email.html) {
    throw new Error('Email must have either text or html content');
  }
  
  // Validate email formats
  const emailRegex = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
  const fullAgentEmail = `${email.agentLocalPart}@${email.agentDomain}`;
  
  if (!emailRegex.test(fullAgentEmail)) {
    throw new Error(`Invalid agent email format: ${fullAgentEmail}`);
  }
  
  if (!emailRegex.test(email.fromEmail)) {
    throw new Error(`Invalid sender email format: ${email.fromEmail}`);
  }
  
  // Validate Message-ID format
  if (!email.messageId.includes('@') || !email.messageId.startsWith('<') || !email.messageId.endsWith('>')) {
    throw new Error(`Invalid Message-ID format: ${email.messageId}`);
  }
}

/**
 * Clean and sanitize email content
 */
export function sanitizeContent(content: string): string {
  if (!content || typeof content !== 'string') {
    return '';
  }
  
  return content
    .trim()
    .replace(/\r\n/g, '\n')        // Normalize line endings
    .replace(/[ \t]+$/gm, '')        // Trim trailing whitespace per line
    .replace(/\n{3,}/g, '\n\n')    // Collapse excessive newlines
    .substring(0, 50000);          // Reasonable content limit
}
