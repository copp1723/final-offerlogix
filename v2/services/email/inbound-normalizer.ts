/**
 * Inbound Email Normalizer
 *
 * Normalizes Mailgun webhook payloads into consistent InboundEmail format.
 * Handles both JSON fields and MIME header parsing.
 */

import type { InboundEmail } from './types';

// ============================================================================
// MAILGUN PAYLOAD INTERFACES
// ============================================================================

interface MailgunWebhookBody {
  // Core fields (always present)
  recipient: string;              // "riley@kunesmacomb.kunesauto.vip"
  sender: string;                 // "lead@example.com"
  subject: string;                // Email subject

  // Content fields
  'body-plain'?: string;          // Plain text body
  'body-html'?: string;           // HTML body

  // Headers (two possible formats)
  'message-headers'?: Array<[string, string]>; // [["Message-Id", "value"], ...]
  'Message-Id'?: string;          // Direct field access
  'In-Reply-To'?: string;         // Direct field access
  'References'?: string;          // Direct field access

  // Alternative content fields
  'stripped-text'?: string;       // Mailgun stripped text
  'stripped-html'?: string;       // Mailgun stripped HTML

  // Webhook metadata
  timestamp?: number;
  token?: string;
  signature?: string;

  // Allow additional fields
  [key: string]: any;
}

/**
 * Parse email address from various formats
 * Handles: "user@domain.com", "Name <user@domain.com>", etc.
 */
function parseEmailAddress(input: string): { email: string; name?: string } {
  if (!input || typeof input !== 'string') {
    throw new Error('Invalid email input');
  }

  const trimmed = input.trim();

  // Format: "Name <user@domain.com>"
  const nameEmailMatch = trimmed.match(/^(.+?)\s*<([^>]+)>$/);
  if (nameEmailMatch) {
    const [, name, email] = nameEmailMatch;
    return {
      email: email.toLowerCase().trim(),
      name: name.trim().replace(/^["']|["']$/g, '') // Remove quotes
    };
  }

  // Format: "user@domain.com"
  const emailMatch = trimmed.match(/^([^@\s]+@[^@\s]+\.[^@\s]+)$/);
  if (emailMatch) {
    return {
      email: emailMatch[1].toLowerCase().trim()
    };
  }

  throw new Error(`Invalid email format: ${input}`);
}

/**
 * Parse domain and local part from email address
 */
function parseEmailParts(email: string): { localPart: string; domain: string } {
  const normalized = email.toLowerCase().trim();
  const atIndex = normalized.lastIndexOf('@');

  if (atIndex === -1) {
    throw new Error(`Invalid email format: ${email}`);
  }

  return {
    localPart: normalized.substring(0, atIndex),
    domain: normalized.substring(atIndex + 1)
  };
}

// ============================================================================
// HEADER PARSING
// ============================================================================

/**
 * Extract headers from Mailgun payload (handles both formats)
 */
function extractHeaders(body: MailgunWebhookBody): Record<string, string> {
  const headers: Record<string, string> = {};

  // Format 1: message-headers array
  const rawHeaders = (body as any)['message-headers'] ?? (body as any)['messageHeaders'];
  let headerPairs: Array<[string, string]> | null = null;
  if (Array.isArray(rawHeaders)) {
    headerPairs = rawHeaders as Array<[string, string]>;
  } else if (typeof rawHeaders === 'string') {
    try {
      const parsed = JSON.parse(rawHeaders);
      if (Array.isArray(parsed)) headerPairs = parsed as Array<[string, string]>;
    } catch {
      // ignore parse error; will rely on direct fields
    }
  }
  if (Array.isArray(headerPairs)) {
    for (const [key, value] of headerPairs) {
      if (typeof key === 'string' && typeof value === 'string') {
        headers[key.toLowerCase()] = value;
      }
    }
  }

  // Format 2: Direct field access (override array values)
  const directHeaders = [
    'Message-Id',
    'In-Reply-To',
    'References',
    'Date',
    'Content-Type'
  ];

  for (const headerName of directHeaders) {
    const value = body[headerName as keyof MailgunWebhookBody];
    if (typeof value === 'string') {
      headers[headerName.toLowerCase()] = value;
    }
  }

  return headers;
}

/**
 * Normalize Message-ID format
 */
function normalizeMessageId(messageId: string): string {
  if (!messageId || typeof messageId !== 'string') {
    throw new Error('Invalid Message-ID');
  }

  const trimmed = messageId.trim();

  // Ensure angle brackets
  if (trimmed.startsWith('<') && trimmed.endsWith('>')) {
    return trimmed;
  }

  return `<${trimmed}>`;
}

/**
 * Parse References header into array of Message-IDs
 */
function parseReferences(references?: string): string[] {
  if (!references || typeof references !== 'string') {
    return [];
  }

  // Split on whitespace and angle brackets, filter empty strings
  const refs = references
    .split(/[\s<>]+/)
    .filter(Boolean)
    .map(ref => ref.includes('@') ? normalizeMessageId(ref) : null)
    .filter(Boolean) as string[];

  return refs;
}

/**
 * Determine threading context from headers
 */
function determineThreadingContext(headers: Record<string, string>, references: string[]): {
  finalInReplyTo: string | null;
  isReply: boolean;
} {
  const inReplyTo = headers['in-reply-to'];
  const hasReferences = references.length > 0;

  // If we have In-Reply-To, use it
  if (inReplyTo) {
    return {
      finalInReplyTo: normalizeMessageId(inReplyTo),
      isReply: true
    };
  }

  // If we have References but no In-Reply-To, use the last Reference
  if (hasReferences) {
    return {
      finalInReplyTo: references[references.length - 1],
      isReply: true
    };
  }

  // No threading context
  return {
    finalInReplyTo: null,
    isReply: false
  };
}

/**
 * Normalize Mailgun webhook payload into InboundEmail
 */
export function normalizeMailgun(body: any): InboundEmail {
  if (!body || typeof body !== 'object') {
    throw new Error('Invalid webhook body');
  }

  const mailgunBody = body as MailgunWebhookBody;

  // Extract and validate required fields
  if (!mailgunBody.recipient || !mailgunBody.sender || !mailgunBody.subject) {
    throw new Error('Missing required fields: recipient, sender, or subject');
  }

  try {
    // Parse recipient (agent) email
    const recipientInfo = parseEmailAddress(mailgunBody.recipient);
    const agentParts = parseEmailParts(recipientInfo.email);

    // Parse sender (lead) email
    const senderInfo = parseEmailAddress(mailgunBody.sender);

    // Extract headers
    const headers = extractHeaders(mailgunBody);

    // Get Message-ID (required)
    const messageIdRaw = headers['message-id'] || mailgunBody['Message-Id'];
    if (!messageIdRaw) {
      throw new Error('Missing Message-ID in email payload');
    }
    const messageId = normalizeMessageId(messageIdRaw);

    // Parse In-Reply-To
    const inReplyToRaw = headers['in-reply-to'] || mailgunBody['In-Reply-To'];

    // Parse References chain
    const referencesRaw = headers['references'] || mailgunBody['References'];
    const references = parseReferences(referencesRaw);

    // Determine final threading context
    const headersWithInReplyTo = { ...headers, 'in-reply-to': inReplyToRaw || '' };
    const { finalInReplyTo } = determineThreadingContext(headersWithInReplyTo, references);

    // Extract content (prefer stripped versions, fallback to full)
    const text = mailgunBody['stripped-text'] || mailgunBody['body-plain'];
    const html = mailgunBody['stripped-html'] || mailgunBody['body-html'];

    // Validate we have some content
    if (!text && !html) {
      console.warn('No content found in email payload');
    }

    // Build normalized result
    const normalized: InboundEmail = {
      agentLocalPart: agentParts.localPart,
      agentDomain: agentParts.domain,
      fromEmail: senderInfo.email, // Already lowercased in parseEmailAddress
      subject: mailgunBody.subject.trim(),
      text: text?.trim(),
      html: html?.trim(),
      messageId,
      inReplyTo: finalInReplyTo,
      references,
      rawHeaders: headers
    };

    return normalized;

  } catch (error) {
    throw new Error(
      `Failed to normalize Mailgun payload: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

// ============================================================================
// VALIDATION HELPERS
// ============================================================================

/**
 * Validate normalized InboundEmail
 */
export function validateInboundEmail(email: InboundEmail): void {
  const required = ['agentLocalPart', 'agentDomain', 'fromEmail', 'subject', 'messageId'];

  for (const field of required) {
    if (!email[field as keyof InboundEmail]) {
      throw new Error(`Missing required field: ${field}`);
    }
  }

  // Validate email formats
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(email.fromEmail)) {
    throw new Error(`Invalid fromEmail format: ${email.fromEmail}`);
  }

  const agentEmail = `${email.agentLocalPart}@${email.agentDomain}`;
  if (!emailRegex.test(agentEmail)) {
    throw new Error(`Invalid agent email format: ${agentEmail}`);
  }

  // Validate Message-ID format
  if (!email.messageId.startsWith('<') || !email.messageId.endsWith('>')) {
    throw new Error(`Invalid Message-ID format: ${email.messageId}`);
  }
}

/**
 * Clean and sanitize email content
 */
export function sanitizeContent(content: string): string {
  if (!content || typeof content !== 'string') {
    return '';
  }

  return content
    .trim()
    .replace(/\r\n/g, '\n')        // Normalize line endings
    .replace(/[ \t]+$/gm, '')        // Trim trailing whitespace per line
    .replace(/\n{3,}/g, '\n\n')    // Collapse excessive newlines
    .substring(0, 50000);          // Reasonable content limit
}
