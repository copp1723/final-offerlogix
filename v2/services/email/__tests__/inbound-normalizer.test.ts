/**
 * Inbound Email Normalizer Tests
 * 
 * Tests for Mailgun payload normalization with various formats
 * and edge cases.
 */

import { describe, test, expect } from '@jest/globals';
import { normalizeMailgun } from '../inbound-normalizer';

// ============================================================================
// TEST DATA
// ============================================================================

const baseMailgunPayload = {
  'message-id': '<test-message-123@example.com>',
  from: 'Customer Name <customer@example.com>',
  to: 'Riley Donovan <riley@kunesmacomb.kunesauto.vip>',
  subject: 'Test Email Subject',
  'body-plain': 'This is the plain text content.',
  'body-html': '<p>This is the HTML content.</p>',
};

const headerArrayPayload = {
  'message-headers': [
    { name: 'Message-ID', value: '<header-msg-456@example.com>' },
    { name: 'From', value: 'Header Customer <header@example.com>' },
    { name: 'To', value: 'Agent Smith <agent@testdomain.com>' },
    { name: 'Subject', value: 'Header Subject' },
    { name: 'In-Reply-To', value: '<previous-msg@example.com>' },
    { name: 'References', value: '<msg1@example.com> <msg2@example.com>' },
    { name: 'Date', value: 'Mon, 01 Jan 2024 12:00:00 +0000' },
  ],
  'body-plain': 'Header array content.',
  'body-html': '<p>Header array HTML.</p>',
};

const threadedPayload = {
  ...baseMailgunPayload,
  'in-reply-to': '<previous-message@example.com>',
  references: '<msg1@example.com> <msg2@example.com> <previous-message@example.com>',
};

const strippedContentPayload = {
  ...baseMailgunPayload,
  'stripped-text': 'Stripped plain text content.',
  'stripped-html': '<p>Stripped HTML content.</p>',
};

// ============================================================================
// TESTS
// ============================================================================

describe('normalizeMailgun', () => {
  
  describe('Basic Normalization', () => {
    test('normalizes basic Mailgun payload', () => {
      const result = normalizeMailgun(baseMailgunPayload);
      
      expect(result).toMatchSnapshot();
      expect(result.agentLocalPart).toBe('riley');
      expect(result.agentDomain).toBe('kunesmacomb.kunesauto.vip');
      expect(result.fromEmail).toBe('customer@example.com');
      expect(result.subject).toBe('Test Email Subject');
      expect(result.messageId).toBe('<test-message-123@example.com>');
      expect(result.text).toBe('This is the plain text content.');
      expect(result.html).toBe('<p>This is the HTML content.</p>');
      expect(result.inReplyTo).toBeNull();
      expect(result.references).toEqual([]);
    });

    test('normalizes payload with message-headers array', () => {
      const result = normalizeMailgun(headerArrayPayload);
      
      expect(result).toMatchSnapshot();
      expect(result.agentLocalPart).toBe('agent');
      expect(result.agentDomain).toBe('testdomain.com');
      expect(result.fromEmail).toBe('header@example.com');
      expect(result.subject).toBe('Header Subject');
      expect(result.messageId).toBe('<header-msg-456@example.com>');
      expect(result.inReplyTo).toBe('<previous-msg@example.com>');
      expect(result.references).toEqual(['<msg1@example.com>', '<msg2@example.com>']);
    });
  });

  describe('Threading Support', () => {
    test('handles threaded email with In-Reply-To and References', () => {
      const result = normalizeMailgun(threadedPayload);
      
      expect(result.inReplyTo).toBe('<previous-message@example.com>');
      expect(result.references).toEqual([
        '<msg1@example.com>',
        '<msg2@example.com>',
        '<previous-message@example.com>'
      ]);
    });

    test('handles empty references', () => {
      const payload = {
        ...baseMailgunPayload,
        references: '',
      };
      
      const result = normalizeMailgun(payload);
      expect(result.references).toEqual([]);
    });

    test('handles comma-separated references', () => {
      const payload = {
        ...baseMailgunPayload,
        references: '<msg1@example.com>,<msg2@example.com>, <msg3@example.com>',
      };
      
      const result = normalizeMailgun(payload);
      expect(result.references).toEqual([
        '<msg1@example.com>',
        '<msg2@example.com>',
        '<msg3@example.com>'
      ]);
    });
  });

  describe('Content Handling', () => {
    test('prefers stripped content over full content', () => {
      const result = normalizeMailgun(strippedContentPayload);
      
      expect(result.text).toBe('Stripped plain text content.');
      expect(result.html).toBe('<p>Stripped HTML content.</p>');
    });

    test('falls back to full content when stripped not available', () => {
      const payload = {
        ...baseMailgunPayload,
        'stripped-text': undefined,
        'stripped-html': undefined,
      };
      
      const result = normalizeMailgun(payload);
      
      expect(result.text).toBe('This is the plain text content.');
      expect(result.html).toBe('<p>This is the HTML content.</p>');
    });

    test('handles missing content gracefully', () => {
      const payload = {
        ...baseMailgunPayload,
        'body-plain': undefined,
        'body-html': undefined,
        'stripped-text': undefined,
        'stripped-html': undefined,
      };
      
      const result = normalizeMailgun(payload);
      
      expect(result.text).toBeUndefined();
      expect(result.html).toBeUndefined();
    });
  });

  describe('Email Address Parsing', () => {
    test('parses email from "Name <email>" format', () => {
      const payload = {
        ...baseMailgunPayload,
        from: 'John Doe <john.doe@example.com>',
        to: 'Agent Smith <agent.smith@company.com>',
      };
      
      const result = normalizeMailgun(payload);
      
      expect(result.fromEmail).toBe('john.doe@example.com');
      expect(result.agentLocalPart).toBe('agent.smith');
      expect(result.agentDomain).toBe('company.com');
    });

    test('handles plain email addresses without angle brackets', () => {
      const payload = {
        ...baseMailgunPayload,
        from: 'plain@example.com',
        to: 'agent@company.com',
      };
      
      const result = normalizeMailgun(payload);
      
      expect(result.fromEmail).toBe('plain@example.com');
      expect(result.agentLocalPart).toBe('agent');
      expect(result.agentDomain).toBe('company.com');
    });

    test('normalizes email addresses to lowercase', () => {
      const payload = {
        ...baseMailgunPayload,
        from: 'Customer <CUSTOMER@EXAMPLE.COM>',
        to: 'Agent <AGENT@COMPANY.COM>',
      };
      
      const result = normalizeMailgun(payload);
      
      expect(result.fromEmail).toBe('customer@example.com');
      expect(result.agentLocalPart).toBe('agent');
      expect(result.agentDomain).toBe('company.com');
    });
  });

  describe('Raw Headers', () => {
    test('builds raw headers from direct fields and message-headers', () => {
      const result = normalizeMailgun(headerArrayPayload);
      
      expect(result.rawHeaders).toMatchObject({
        'message-id': '<header-msg-456@example.com>',
        'from': 'Header Customer <header@example.com>',
        'to': 'Agent Smith <agent@testdomain.com>',
        'subject': 'Header Subject',
        'in-reply-to': '<previous-msg@example.com>',
        'references': '<msg1@example.com> <msg2@example.com>',
        'date': 'Mon, 01 Jan 2024 12:00:00 +0000',
      });
    });

    test('includes direct fields in raw headers', () => {
      const result = normalizeMailgun(baseMailgunPayload);
      
      expect(result.rawHeaders).toMatchObject({
        'message-id': '<test-message-123@example.com>',
        'from': 'Customer Name <customer@example.com>',
        'to': 'Riley Donovan <riley@kunesmacomb.kunesauto.vip>',
        'subject': 'Test Email Subject',
      });
    });
  });

  describe('Error Handling', () => {
    test('throws error for missing Message-ID', () => {
      const payload = {
        ...baseMailgunPayload,
        'message-id': undefined,
      };
      
      expect(() => normalizeMailgun(payload)).toThrow('Missing required Message-ID header');
    });

    test('throws error for missing From header', () => {
      const payload = {
        ...baseMailgunPayload,
        from: undefined,
      };
      
      expect(() => normalizeMailgun(payload)).toThrow('Missing required From header');
    });

    test('throws error for missing To header', () => {
      const payload = {
        ...baseMailgunPayload,
        to: undefined,
      };
      
      expect(() => normalizeMailgun(payload)).toThrow('Missing required To header');
    });

    test('throws error for missing Subject header', () => {
      const payload = {
        ...baseMailgunPayload,
        subject: undefined,
      };
      
      expect(() => normalizeMailgun(payload)).toThrow('Missing required Subject header');
    });

    test('throws error for invalid To header format', () => {
      const payload = {
        ...baseMailgunPayload,
        to: 'invalid-email-format',
      };
      
      expect(() => normalizeMailgun(payload)).toThrow('Invalid To header format');
    });
  });

  describe('Alternative Field Names', () => {
    test('uses sender field as fallback for from', () => {
      const payload = {
        ...baseMailgunPayload,
        from: undefined,
        sender: 'Sender Name <sender@example.com>',
      };
      
      const result = normalizeMailgun(payload);
      expect(result.fromEmail).toBe('sender@example.com');
    });

    test('uses recipient field as fallback for to', () => {
      const payload = {
        ...baseMailgunPayload,
        to: undefined,
        recipient: 'Recipient <recipient@company.com>',
      };
      
      const result = normalizeMailgun(payload);
      expect(result.agentLocalPart).toBe('recipient');
      expect(result.agentDomain).toBe('company.com');
    });
  });
});
