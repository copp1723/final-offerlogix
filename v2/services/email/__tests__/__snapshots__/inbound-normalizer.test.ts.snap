// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`normalizeMailgun Basic Normalization normalizes basic Mailgun payload 1`] = `
{
  "agentDomain": "kunesmacomb.kunesauto.vip",
  "agentLocalPart": "riley",
  "fromEmail": "customer@example.com",
  "html": "<p>This is the HTML content.</p>",
  "inReplyTo": null,
  "messageId": "<test-message-123@example.com>",
  "rawHeaders": {
    "from": "Customer Name <customer@example.com>",
    "message-id": "<test-message-123@example.com>",
    "subject": "Test Email Subject",
    "to": "Riley Donovan <riley@kunesmacomb.kunesauto.vip>",
  },
  "references": [],
  "subject": "Test Email Subject",
  "text": "This is the plain text content.",
}
`;

exports[`normalizeMailgun Basic Normalization normalizes payload with message-headers array 1`] = `
{
  "agentDomain": "testdomain.com",
  "agentLocalPart": "agent",
  "fromEmail": "header@example.com",
  "html": "<p>Header array HTML.</p>",
  "inReplyTo": "<previous-msg@example.com>",
  "messageId": "<header-msg-456@example.com>",
  "rawHeaders": {
    "date": "Mon, 01 Jan 2024 12:00:00 +0000",
    "from": "Header Customer <header@example.com>",
    "in-reply-to": "<previous-msg@example.com>",
    "message-id": "<header-msg-456@example.com>",
    "references": "<msg1@example.com> <msg2@example.com>",
    "subject": "Header Subject",
    "to": "Agent Smith <agent@testdomain.com>",
  },
  "references": [
    "<msg1@example.com>",
    "<msg2@example.com>",
  ],
  "subject": "Header Subject",
  "text": "Header array content.",
}
`;
